File: src/main/java/com/dbn/mybatis/plugins/EnumsPlugin.java
package com.dbn.mybatis.plugins;

import com.dbn.mybatis.DbnMyBatisGenerator;
import com.dbn.mybatis.model.Config;
import com.dbn.mybatis.model.EnumMustache;
import com.dbn.mybatis.model.EnumMustacheField;
import com.dbn.utils.StringUtil;
import com.github.mustachejava.DefaultMustacheFactory;
import com.github.mustachejava.Mustache;
import com.github.mustachejava.MustacheFactory;
import com.intellij.openapi.util.io.StreamUtil;
import lombok.Cleanup;
import lombok.SneakyThrows;
import lombok.val;
import org.mybatis.generator.api.IntrospectedColumn;
import org.mybatis.generator.api.IntrospectedTable;
import org.mybatis.generator.api.PluginAdapter;
import org.mybatis.generator.api.dom.java.Field;
import org.mybatis.generator.api.dom.java.TopLevelClass;
import org.mybatis.generator.config.Context;
import org.mybatis.generator.internal.DefaultShellCallback;

import java.io.File;
import java.io.InputStream;
import java.io.PrintWriter;
import java.io.StringReader;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class EnumsPlugin extends PluginAdapter {
    private static final String ENUM_FLAG = "枚举";
    private final MustacheFactory mf = new DefaultMustacheFactory();
    private Config config;

    @Override
    public void setContext(Context context) {
        config = DbnMyBatisGenerator.getConfig();
        super.setContext(context);
    }

    @Override
    public boolean validate(List<String> warnings) {
        return true;
    }

    @SneakyThrows
    @Override
    public boolean modelFieldGenerated(Field field, TopLevelClass topLevelClass,
                                       IntrospectedColumn introspectedColumn,
                                       IntrospectedTable introspectedTable,
                                       ModelClassType modelClassType) {
        val remark = introspectedColumn.getRemarks();

        if (remark.contains(ENUM_FLAG)) {
            generateEnum(field, remark);
        }

        return super.modelFieldGenerated(field, topLevelClass, introspectedColumn,
                introspectedTable, modelClassType);
    }

    private void generateEnum(Field field, String remark) throws Exception {
        String name = StringUtil.toCamelCase(field.getName());
        name = StringUtil.capitalizeFirstWord2(name);
        String enumClassName = name + config.getEnumPostfixPackage();
        String enumValue = remark.split("-")[1];
        String[] enums = enumValue.split(";");

        List<EnumMustacheField> list = Arrays.stream(enums)
                .map(s -> {
                    val tmpList = s.split(":");
                    val enumMustacheField = new EnumMustacheField();
                    enumMustacheField.label = tmpList[0];
                    enumMustacheField.setCode(Integer.parseInt(tmpList[1]));
                    enumMustacheField.msg = tmpList[2];
                    return enumMustacheField;
                }).collect(Collectors.toList());

        EnumMustache enumMustache = new EnumMustache();
        enumMustache.enumPath = config.getEnumPackage();
        enumMustache.enumClassName = enumClassName;
        enumMustache.enumFields = list;

        String targetProject = config.getProjectFolder() + "/" + config.getEnumMvnPath();
        File directory = new DefaultShellCallback(true).getDirectory(targetProject, config.getEnumPackage());
        File targetFile = new File(directory, enumClassName + ".java");

        URL url = getClass().getClassLoader().getResource("mybatis/Enum.mustache");
        @Cleanup
        @SuppressWarnings("DataFlowIssue")
        InputStream inputStream = url.openStream();
        String str = new String(StreamUtil.readBytes(inputStream), StandardCharsets.UTF_8);
        Mustache mustache = mf.compile(new StringReader(str), "Enum.mustache");

        mustache.execute(new PrintWriter(targetFile), enumMustache).flush();
    }
}

File: src/main/kotlin/com/dbn/mybatis/plugin/EnumsPlugin.kt
package com.dbn.mybatis.plugin

import com.dbn.mybatis.model.EnumMustache
import com.dbn.mybatis.model.EnumMustacheField
import com.dbn.utils.StringUtil
import com.github.mustachejava.DefaultMustacheFactory
import com.github.mustachejava.MustacheFactory
import org.mybatis.generator.api.IntrospectedColumn
import org.mybatis.generator.api.IntrospectedTable
import org.mybatis.generator.api.Plugin.ModelClassType
import org.mybatis.generator.api.PluginAdapter
import org.mybatis.generator.api.dom.java.Field
import org.mybatis.generator.api.dom.java.TopLevelClass
import org.mybatis.generator.config.Context
import org.mybatis.generator.internal.DefaultShellCallback
import java.io.File
import java.io.PrintWriter
import java.io.StringReader
import java.nio.charset.StandardCharsets

/**
 * 根据注释生成枚举类,注释格式必须满足如下所示:<br>
 * 性别,枚举-MALE:1:男;FEMALE:2:女
 *
 * @author yudong
 */
class EnumsPlugin : PluginAdapter() {
    private val mf: MustacheFactory = DefaultMustacheFactory()

    private lateinit var enumPath: String
    private lateinit var targetProject: String

    override fun setContext(context: Context) {
        enumPath = ""
        targetProject = context.javaModelGeneratorConfiguration.targetProject
    }

    override fun validate(warnings: List<String>): Boolean {
        return true
    }

    override fun modelFieldGenerated(
        field: Field, topLevelClass: TopLevelClass, introspectedColumn: IntrospectedColumn,
        introspectedTable: IntrospectedTable, modelClassType: ModelClassType,
    ): Boolean {
        val remark = introspectedColumn.remarks

        if (remark.contains(ENUM_FLAG)) {
            generateEnum(field, remark)
        }

        return super.modelFieldGenerated(field, topLevelClass, introspectedColumn, introspectedTable, modelClassType)
    }

    private fun generateEnum(field: Field, remark: String) {
        // 性别,枚举-MALE:1:男;FEMALE:2:女
        val enumClassName = StringUtil.toCamelCase(field.name) + "Enum"
        val enumValue = remark.split("-")[1]
        val enums = enumValue.split(";")

        val list = enums
            .map { s: String ->
                val tmpList = s.split(":")
                val enumMustacheField = EnumMustacheField()
                enumMustacheField.label = tmpList[0]
                enumMustacheField.code = tmpList[1].toInt()
                enumMustacheField.msg = tmpList[2]
                enumMustacheField
            }

        val enumMustache = EnumMustache()
        enumMustache.enumPath = enumPath
        enumMustache.enumClassName = enumClassName
        enumMustache.enumFields = list

        val directory = DefaultShellCallback(true).getDirectory(targetProject, enumPath)
        val targetFile = File(directory, "$enumClassName.java")

        val url = javaClass.classLoader.getResource("mybatis/Enum.mustache")
        val str = url.readText(StandardCharsets.UTF_8)
        val mustache = mf.compile(StringReader(str), "Enum.mustache")

        mustache.execute(PrintWriter(targetFile), enumMustache).flush()
    }


    companion object {
        private const val ENUM_FLAG = "枚举"
    }
}
